grovedb/src/operations/delete.rs

```diff
@@ -24,24 +81,38 @@ impl GroveDb {
             ))
         } else {
             let element = self.get_raw(path_iter.clone(), key.as_ref(), transaction)?;
-            {
-                let (mut merk, prefix) = self.get_subtrees().get(path_iter.clone(), transaction)?;
-                Element::delete(&mut merk, key, transaction)?;
-
-                // after deletion, if there is a transaction, add the merk back into the hashmap
+            let subtrees = self.get_subtrees();
+            let delete_element = || -> Result<(), Error> {
+                // TODO: we shouldn't handle this context manually each time
+                let (mut parent_merk, prefix) = subtrees.get(path_iter.clone(), transaction)?;
+                Element::delete(&mut parent_merk, &key, transaction)?;
                 if let Some(prefix) = prefix {
-                    self.get_subtrees()
-                        .insert_temp_tree_with_prefix(prefix, merk, transaction);
+                    subtrees.insert_temp_tree_with_prefix(prefix, parent_merk, transaction);
                 } else {
-                    self.get_subtrees()
-                        .insert_temp_tree(path_iter.clone(), merk, transaction);
+                    subtrees.insert_temp_tree(path_iter.clone(), parent_merk, transaction);
                 }
-            }
+                Ok(())
+            };
 
             if let Element::Tree(_) = element {
+                let subtree_merk_path = path_iter.clone().chain(std::iter::once(key));
+                let subtrees_paths = self.find_subtrees(subtree_merk_path.clone(), transaction)?;
+                let (subtree_merk, prefix) = subtrees.get(subtree_merk_path, transaction)?;
+
+                let is_empty = subtree_merk.is_empty_tree(transaction);
+
+                if let Some(prefix) = prefix {
+                    subtrees.insert_temp_tree_with_prefix(prefix, subtree_merk, transaction);
+                } else {
+                    subtrees.insert_temp_tree(path_iter.clone(), subtree_merk, transaction);
+                }
+
+                if only_delete_tree_if_empty && !is_empty {
```

Comment:
same here, is this ordering important?  Why not
```suggestion
                if let Some(prefix) = prefix {
                    subtrees.insert_temp_tree_with_prefix(prefix, subtree_merk, transaction);
                } else {
                    subtrees.insert_temp_tree(path_iter.clone(), subtree_merk, transaction);
                }

                if only_delete_tree_if_empty && !subtree_merk.is_empty_tree(transaction) {
```